tools/AduCmdlets/create-adu-import-manifest.sh

```diff
@@ -0,0 +1,289 @@
+        fi
+
+        INSTALLED_CRITERIA=$OPTARG
+
+        # POSIX regex only, sigh.
+        if ! [[ "$INSTALLED_CRITERIA" =~ ^[^[:space:]]{1,64}$ ]]; then
+            write_error "Invalid installed criteria specified."
+            exit 1
+        fi
+        ;;
+
+    n)
+        if [ ! -z "$UPDATE_NAME" ]; then
+            write_error "Update name specified twice."
+            exit 1
+        fi
+
+        UPDATE_NAME=$OPTARG
+
+        if ! [[ "$UPDATE_NAME" =~ ^[a-zA-Z0-9.-]{1,64}$ ]]; then
+            write_error "Invalid update name specified."
+            exit 1
+        fi
+        ;;
+    p)
+        if [ ! -z "$PROVIDER_NAME" ]; then
+            write_error "Provider name specified twice."
+            exit 1
+        fi
+
+        PROVIDER_NAME=$OPTARG
+
+        if ! [[ "$PROVIDER_NAME" =~ ^[a-zA-Z0-9.-]{1,64}$ ]]; then
+            write_error "Invalid provider name specified."
+            exit 1
+        fi
+        ;;
+    t)
+        if [ ! -z "$UPDATE_TYPE" ]; then
+            write_error "Update type specified twice."
+            exit 1
+        fi
+
+        UPDATE_TYPE=$OPTARG
+
+        # POSIX regex only, sigh.
+        if ! [[ "$UPDATE_TYPE" =~ ^[^[:space:]]+/[^[:space:]]+:[[:digit:]]{1,5}$ ]]; then
+            write_error "Invalid update type specified."
+            exit 1
+        fi
+        ;;
+
+    v)
+        if [ ! -z "$UPDATE_VERSION" ]; then
+            write_error "Update version specified twice."
+            exit 1
+        fi
+
+        UPDATE_VERSION=$OPTARG
+        ;;
+    \?)
+        # Unsupported option
+        print_usage
+        exit 1
+        ;;
+    :)
+        # Missing option argument
+        print_usage
+        exit 1
+        ;;
+    *)
+        # Unimplemented option
+        exit 1
+        ;;
+    esac
+done
+
+# Update files are the arguments without switches.
+shift $(expr $OPTIND - 1)
+UPDATE_FILES=($@)
+
+# Verify that all required arguments were specified and correct.
+
+if [ -z "$UPDATE_NAME" ]; then
+    write_error 'Update name not specified.'
+    exit 1
+fi
+
+if [ -z "$UPDATE_TYPE" ]; then
+    write_error 'Update type not specified.'
+    exit 1
+fi
+
+if [ -z "$INSTALLED_CRITERIA" ]; then
+    write_error 'Installed criteria not specified.'
+    exit 1
+fi
+
+if [ -z "$PROVIDER_NAME" ]; then
+    write_error 'Provider name not specified.'
+    exit 1
+fi
+
+if [ -z "$UPDATE_VERSION" ]; then
+    write_error 'Update version not specified.'
+    exit 1
+fi
+
+if [ ${#COMPAT_INFOS[@]} -eq 0 ]; then
+    write_error 'Compatibility info not specified.'
+    exit 1
+fi
+
+if [ ${#UPDATE_FILES[@]} -eq 0 ]; then
+    write_error 'Update file(s) not specified.'
+    exit 1
+fi
+
+for file in "${UPDATE_FILES[@]}"; do
+    if [ ! -f "$file" ]; then
+        write_error "File '$file' not found."
+        exit 1
+    fi
+done
+
+# Match Powershell's .ToString('o') format, e.g. "2021-04-14T18:14:35.5816196Z"
+CREATED_DATETIME=$(date --utc "+%Y-%m-%dT%H:%M:%S.%NZ")
+
```

Comment:
checking with Leo, but `date --utc --iso-8601=seconds` might suffice here if ns aren't significant, and .NET DateTime is being used to parse.